internal/git/git.go

```diff
@@ -14,44 +15,55 @@ const (
 
 type checkoutCmd struct {
 	rukpakv1alpha1.GitSource
+	User  string
+	Token string
 }
 
-func CloneCommandFor(s rukpakv1alpha1.GitSource) (string, error) {
-	cmd := &checkoutCmd{GitSource: s}
+func CloneCommandFor(s rukpakv1alpha1.GitSource, user, token string) (string, error) {
+	cmd := &checkoutCmd{GitSource: s, User: user, Token: token}
 	err := cmd.Validate()
 	if err != nil {
 		return "", err
 	}
-	return cmd.String(), nil
+	return cmd.String()
 }
 
-func (c *checkoutCmd) String() string {
+func (c *checkoutCmd) String() (string, error) {
 	var checkoutCommand string
 	var repository = c.Repository
 	var directory = c.Directory
 	var branch = c.Ref.Branch
 	var commit = c.Ref.Commit
 	var tag = c.Ref.Tag
 
+	if c.User != "" || c.Token != "" {
+		repositoryURL, err := url.Parse(repository)
+		if err != nil {
+			return "", err
+		}
```

Comment:
Git supports several [protocols](https://git-scm.com/book/en/v2/Git-on-the-Server-The-Protocols), (not just http) so it seems like we need to support them or at least have our API make it obvious which of them we support?

I wonder if it would be easier to refactor the git source using a git library before we try to tackle git auth stuff?

Reply:
I kind of agree, the original implementation got us far enough but we probably need to overhaul it to be more robust. I wonder if it makes more sense to have #191 land first so the git codepath can be a lot cleaner. I can open a tracking issue for this and we can see how we prioritize it.

Reply:
What's our story around rate limiting when using git clients under-the-hood then? Instructing users to configure a Secret reference so we can have proper authentication? That seems potentially reasonable, but I'm not sure if that's surprising UX behavior, or whether this introduces a more leaky abstraction.

Reply:
See #291 for tracking issue

Reply:
I would like to implement the authentication with the current design for now.  I'll work on the #291 after that.

Reply:
I'm still a little concerned about the impact of this change on SSH remotes. To me, it seems like this will make the API more brittle (i.e. "you can use a secret with http remotes, but if you try to use it with SSH remotes, things will break in potentially weird ways")

Reply:
I see.  The `url.Parse` should be used only for https protocol.  This PR only works with the https protocol.  May I make the SSH remote case in the follow up issue #294?

Reply:
I'll update to return the `repository` as-is when the url.Parse so that the error is the same one without the secret provided.

Reply:
Thinking about this more, I suppose that ssh remotes have never worked because we don't have anything that supports providing SSH credentials, which are required to even make the SSH connection. So I suppose we're okay on that front.

Should we perhaps improve docs/validation to make it clear that only SSH is supported?

Reply:
I updated the description of the `Secret.Secret`.  Thanks!